actors/builtin/miner/miner_actor.go

```diff
@@ -541,88 +652,172 @@ func (a Actor) TerminateSectors(rt Runtime, params *TerminateSectorsParams) *adt
 	rt.State().Readonly(&st)
 	rt.ValidateImmediateCallerIs(st.Info.Worker)
 
-	sectorNos := bitfieldToSectorNos(rt, params.Sectors)
-
 	// Note: this cannot terminate pre-committed but un-proven sectors.
 	// They must be allowed to expire (and deposit burnt).
-	a.terminateSectors(rt, sectorNos, power.SectorTerminationManual)
+	a.terminateSectors(rt, params.Sectors, power.SectorTerminationManual)
 	return nil
 }
 
 ////////////
 // Faults //
 ////////////
 
-type DeclareTemporaryFaultsParams struct {
-	SectorNumbers abi.BitField
-	Duration      abi.ChainEpoch
+type DeclareFaultsParams struct {
+	Faults []FaultDeclaration
 }
 
-func (a Actor) DeclareTemporaryFaults(rt Runtime, params *DeclareTemporaryFaultsParams) *adt.EmptyValue {
-	if params.Duration <= abi.ChainEpoch(0) {
-		rt.Abortf(exitcode.ErrIllegalArgument, "non-positive fault Duration %v", params.Duration)
+type FaultDeclaration struct {
+	Deadline uint64 // In range [0..WPoStPeriodDeadlines)
+	Sectors  *abi.BitField
+}
+
+func (a Actor) DeclareFaults(rt Runtime, params *DeclareFaultsParams) *adt.EmptyValue {
+	if uint64(len(params.Faults)) > WPoStPeriodDeadlines {
+		rt.Abortf(exitcode.ErrIllegalArgument, "too many declarations %d, max %d", len(params.Faults), WPoStPeriodDeadlines)
 	}
 
-	effectiveEpoch := rt.CurrEpoch() + DeclaredFaultEffectiveDelay
+	currEpoch := rt.CurrEpoch()
+	store := adt.AsStore(rt)
 	var st State
-	requiredFee := rt.State().Transaction(&st, func() interface{} {
+	var declaredFaultSectors []*SectorOnChainInfo
+	var detectedFaultSectors []*SectorOnChainInfo
+	var penalty abi.TokenAmount
+
+	rt.State().Transaction(&st, func() interface{} {
 		rt.ValidateImmediateCallerIs(st.Info.Worker)
 
-		store := adt.AsStore(rt)
-		storageWeightDescs := []*power.SectorStorageWeightDesc{}
-		dfaults, err := params.SectorNumbers.All(MaxFaultsCount)
-		if err != nil {
-			rt.Abortf(exitcode.ErrIllegalArgument, "failed to enumerate faulted sector list")
+		// The proving period start may be negative for low epochs, but all the arithmetic should work out
+		// correctly in order to declare faults for an upcoming deadline or the next period.
+		periodStart, _ := st.ProvingPeriodStart(currEpoch)
+		deadlines, err := LoadDeadlines(adt.AsStore(rt), &st)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadlines")
+
+		// Traverse earlier submissions and enact detected faults.
+		// This is necessary to prevent the miner "declaring" a fault for a PoSt already missed.
+		deadlineIndex, _ := ComputeCurrentDeadline(periodStart, currEpoch)
+		detectedFaultSectors, penalty = checkMissingPoStFaults(rt, st, store, deadlines, periodStart, deadlineIndex, currEpoch)
+
+		var decaredSectors []*abi.BitField
+		for _, decl := range params.Faults {
+			err := validateFRDeclaration(currEpoch, periodStart, deadlines, decl.Deadline, decl.Sectors)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalArgument, "invalid fault declaration")
```

Comment:
error should be more explicit: "Faults for deadline decl.Deadline are too late for submission"

Reply:
The nested `err` will be included in the log and has those details.